EIPS/eip-1193.md

```diff
@@ -12,99 +12,142 @@ requires: 155, 695, 1102, 1474
+Consult each RPC method's documentation for its return type.
+You can find a list of common Ethereum JSON-RPC methods [here](https://github.com/ethereum/wiki/wiki/JSON-RPC#json-rpc-api-reference).
 
-See [available JSON-RPC methods](https://github.com/ethereumproject/go-ethereum/wiki/JSON-RPC#json-rpc-methods).
+The Promise rejects with errors of the following form:
 
-### Events
+```typescript
+{
+  message: string,
+  code: number,
+  data?: any
+}
+```
 
-Events are emitted using [EventEmitter](https://nodejs.org/api/events.html).
+See the [JSON-RPC Errors](#json-rpc-errors) section for more details.
 
-#### notification
+### sendAsync (DEPRECATED)
 
-All subscriptions from the node emit on notification. Attach listeners with:
+Submits a JSON-RPC request to the Provider.
+As [`ethereum.request`](#request), but with a callback and JSON-RPC objects.
 
-```js
-ethereum.on('notification', listener: (result: any) => void): this;
+```typescript
+ethereum.sendAsync(request: Object, callback: Function): Object;
 ```
 
-To create a subscription, call `ethereum.send('eth_subscribe', [])` or `ethereum.send('shh_subscribe', [])`.
+The interfaces of request and response objects are not specified here.
+Historically, they have followed the [Ethereum JSON-RPC specification](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1474.md).
 
-See the [eth subscription methods](https://github.com/ethereum/go-ethereum/wiki/RPC-PUB-SUB#supported-subscriptions) and [shh subscription methods](https://github.com/ethereum/go-ethereum/wiki/Whisper-v6-RPC-API#shh_subscribe).
+### send (DEPRECATED)
 
-#### connect
+Due to conflicting implementations and specifications, this method is unreliable and should not be used.
 
-The provider emits `connect` on connect to a network.
+```typescript
+ethereum.send(...args: Array<any>): unknown;
+```
 
-```js
-ethereum.on('connect', listener: () => void): this;
+### Events
+
+Events follow the [Node.js `EventEmitter`](https://nodejs.org/api/events.html) API.
+
+#### notification
+
+The Provider emits `notification` upon receipt of a JSON-RPC notification intended for the dapp.
+
+```typescript
+interface JsonRpcNotification {
+  method: string,
+  params: Array<unknown> | Object,
+  jsonrpc?: '2.0'
+}
+
+ethereum.on('notification', listener: (notification: JsonRpcNotification) => void): ethereum;
 ```
 
-You can detect which chain by sending `eth_chainId`:
+Notifications include `eth_subscription` messages, if supported by the Provider. See [here](https://geth.ethereum.org/docs/rpc/pubsub) for details.
 
-```js
-const chainId = await ethereum.send('eth_chainId');
-> 1
+#### connect
+
+The Provider emits `connect` when it becomes connected. This includes:
+
+- when the Provider first connects to a chain after being initialized
+- when the Provider connects to a chain, after the `close` event was emitted
+
+```typescript
+ethereum.on('connect', listener: (chainId: number) => void): ethereum;
```

Comment:
Should the callback listener always return a number? Might be better (read: more flexible) to return an object that has the `chainId` instead:

```typescript
interface Network {
  chainId: number
}

ethereum.on('connect', listener: (network: Network) => void): ethereum;
```

This way it's:

1. More extensible should we want to add more stuff;
2. People are immediately clear what the number is about.

Reply:
Great point, I'll add that.

Reply:
Added in 31f7471